mm2src/coins/hd_wallet/mod.rs

```diff
@@ -0,0 +1,431 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+mod account_ops;
+pub use account_ops::HDAccountOps;
+
+mod address_ops;
+pub use address_ops::HDAddressOps;
+
+mod coin_ops;
+pub use coin_ops::{HDAddressId, HDWalletCoinOps};
+
+mod confirm_address;
+#[cfg(test)]
+pub(crate) use confirm_address::for_tests::MockableConfirmAddress;
+pub(crate) use confirm_address::{ConfirmAddressStatus, RpcTaskConfirmAddress};
+pub use confirm_address::{HDConfirmAddress, HDConfirmAddressError};
+
+mod errors;
+pub use errors::{AccountUpdatingError, AddressDerivingError, InvalidBip44ChainError, NewAccountCreatingError,
+                 NewAddressDeriveConfirmError, NewAddressDerivingError};
+
+mod pubkey;
+pub use pubkey::{ExtractExtendedPubkey, HDExtractPubkeyError, HDXPubExtractor, RpcTaskXPubExtractor};
+
+mod storage;
+#[cfg(target_arch = "wasm32")]
+pub(crate) use storage::HDWalletDb;
+#[cfg(test)] pub(crate) use storage::HDWalletMockStorage;
+pub use storage::{HDAccountStorageItem, HDWalletCoinStorage, HDWalletId, HDWalletStorageError, HDWalletStorageOps};
+pub(crate) use storage::{HDWalletStorageInternalOps, HDWalletStorageResult};
+
+mod wallet_ops;
+pub use wallet_ops::HDWalletOps;
+
+pub(crate) type AddressDerivingResult<T> = MmResult<T, AddressDerivingError>;
+pub(crate) type HDAccountsMap<HDAccount> = BTreeMap<u32, HDAccount>;
+pub(crate) type HDAccountsMutex<HDAccount> = AsyncMutex<HDAccountsMap<HDAccount>>;
+pub(crate) type HDAccountsMut<'a, HDAccount> = AsyncMutexGuard<'a, HDAccountsMap<HDAccount>>;
+pub(crate) type HDAccountMut<'a, HDAccount> = AsyncMappedMutexGuard<'a, HDAccountsMap<HDAccount>, HDAccount>;
+pub(crate) type HDWalletAddress<T> =
+    <<<T as HDWalletOps>::HDAccount as HDAccountOps>::HDAddress as HDAddressOps>::Address;
+pub(crate) type HDCoinAddress<T> = HDWalletAddress<<T as HDWalletCoinOps>::HDWallet>;
+pub(crate) type HDWalletHDAddress<T> = <<T as HDWalletOps>::HDAccount as HDAccountOps>::HDAddress;
+pub(crate) type HDCoinHDAddress<T> = HDWalletHDAddress<<T as HDWalletCoinOps>::HDWallet>;
+pub(crate) type HDWalletHDAccount<T> = <T as HDWalletOps>::HDAccount;
+pub(crate) type HDCoinHDAccount<T> = HDWalletHDAccount<<T as HDWalletCoinOps>::HDWallet>;
+
+pub(crate) const DEFAULT_GAP_LIMIT: u32 = 20;
+const DEFAULT_ACCOUNT_LIMIT: u32 = ChildNumber::HARDENED_FLAG;
+const DEFAULT_ADDRESS_LIMIT: u32 = ChildNumber::HARDENED_FLAG;
+const DEFAULT_RECEIVER_CHAIN: Bip44Chain = Bip44Chain::External;
+
+#[derive(Clone)]
+pub struct HDAddress<Address, Pubkey> {
+    pub address: Address,
+    pub pubkey: Pubkey,
+    pub derivation_path: DerivationPath,
+}
+
+impl<Address, Pubkey> HDAddressOps for HDAddress<Address, Pubkey>
+where
+    Address: Clone + Display + Send + Sync,
+{
+    type Address = Address;
+    type Pubkey = Pubkey;
+
+    fn address(&self) -> &Self::Address { &self.address }
+
+    fn pubkey(&self) -> &Self::Pubkey { &self.pubkey }
+
+    fn derivation_path(&self) -> &DerivationPath { &self.derivation_path }
+}
```

Comment:
Thank you for bringing this point up to discussion, I decided to add `HDAddressOps` for 2 reasons:

1 - While the `HDAddress` generic struct serves as a general-purpose representation suitable for almost all coins, some coins in the future might require additional fields. An example would be `ZCoin` where it might be good to implement this trait for a slightly different `ZcoinHDAddress` Struct that includes and additional field for `my_z_addr_encoded`. https://github.com/KomodoPlatform/komodo-defi-framework/blob/6b751b865b7a3020471fd06ce4e80c6d1ccffc65/mm2src/coins/z_coin.rs#L202
Zcoin will probably need a different struct from the generic `HDAccount` too depending how we handle this case https://github.com/KomodoPlatform/komodo-defi-framework/pull/1933#discussion_r1293013105, but the generic `HDWallet` will probably be reused for `Zcoin` where we plug `ZcoinHDAccount` into it.

2 - Another reason was to design a crate for the framework/SDK, SDKs crates best practices often involve providing traits with a generic struct that implements these traits, this allows the users of such crates to opt out of using the generic struct if they want, an example would be a user that wants to only keep the pubkey inside their custom `HDAddress` struct and use `fn address(&self) -> &Self::Address` to derive the address from the pubkey.

I am not completely against removing `HDAddressOps` but wanted to provide some context into the thinking process that went into it, if you still think it's good to remove it to reduce boilerplate code I won't object :)

Reply:
These are great points, so I don't think that it's good to remove this and other related traits anymore, thanks 🙂 

I have one note regarding
> use fn address(&self) -> &Self::Address to derive the address from the pubkey.

Since the method returns reference, I think it won't be possible to return the derived address because the value will exist only in the method body.

Reply:
> Since the method returns reference, I think it won't be possible to return the derived address because the value will exist only in the method body.

Fixed the method to return a non-reference.